### PR TITLE
fix #13443 Underscore cannot be defined more than once in proc

### LINKED ISSUE #13443: Underscore cannot be defined more than once in proc
**Test**
```nim
proc foo(_, _: int) = discard
```

**Expected**
Compiles with no issue. It works for one underscore, though I'm not sure this is even a feature.

**Got**
```nim
Test.nim(1, 13) Error: attempt to redefine: '_'
```
***
```nim
>nim -v
Nim Compiler Version 1.1.1 [Linux: i386]
Compiled at 2020-02-19
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 8c22518d673cf792fa064d2d561d1c6f03b823bb
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ringabout**
fix #13443

## future work
- [ ] make it work for js+vm  (see disabled test)
- [ ] add test for:
```nim
template main =
  proc foo(_: int) =
    let a = _
doAssert not compiles(foo(1))
```
- [ ] this doesn't work inside templates:

```nim
import std/sequtils

proc main() =
  iterator fn(_, _: int, c: int): int = yield c
  doAssert toSeq(fn(1,2,3)) == @[3]
static: main() # ok
main() # ok

template main2() =
  iterator fn(_, _: int, c: int): int = yield c
main2() # bug: Error: attempt to redefine: '_`gensym3'
```
ditto with procs:
```
when true:
  template main() =
    proc fn(_, _, a: int): auto =
      a
    echo fn(1,2,3) # Error: attempt to redefine: '_`gensym0'
  main()
```

- [ ] this works but shouldn't:
```nim
when defined case5:
  template main() =
    proc fn(_, a: int): auto =
      echo _
      a
    echo fn(1,2)
  main()
```

- [ ] this is a (desirable) breaking change but should be documented in changelog:
```nim
when defined case3:
  proc fn(_: int, a: int): auto =
    echo _
    a
  echo fn(1,2)
```
before PR: works
after PR: Error: undeclared identifier: '_'

- [ ] see whether we should add `__attribute__((unused))` in cgen for the corresponding parameters

### CODE REVIEW

File: tests/proc/tunderscore.nim [hunk 1]
```diff
@@ -0,0 +1,28 @@
+discard """
+  targets: "c cpp js"
+"""
+
```
Comment (timotheecour) [inline tests/proc/tunderscore.nim:36]: does it also help with https://github.com/nim-lang/Nim/issues/13804 ?

**Comment by ringabout** [comment]
It doesn't work with templates.
Hi @timotheecour do you think I should support it? Or some suggestions regarding it?
```nim
template main() =
  var x = 0
  proc foo(_, _: int) = x += 5

  foo(1, 2)
  doAssert x == 5
```

**Comment by timotheecour** [comment]
> do you think I should support it?

yes i think that snippet should compile. can be done in subsequent PR though

File: tests/proc/tunderscore.nim (hunk 1 above)
Comment (ringabout) [reply tests/proc/tunderscore.nim:36]: no

File: compiler/semtypes.nim [hunk 2]
```diff
@@ -40,6 +40,12 @@ const
   errNoGenericParamsAllowedForX = "no generic parameters allowed for $1"
   errInOutFlagNotExtern = "the '$1' modifier can be used only with imported types"
 
+
+proc isDiscardUnderscoreV(v: PSym): bool =
+  if v.name.s == "_":
+    v.flags.incl(sfGenSym)
```
Comment (Clyybber) [inline compiler/semtypes.nim]: Don't hide mutation in a proc with an `isX` name. I would just inline this proc.

File: tests/proc/tunderscore.nim [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
```
Comment (timotheecour) [inline tests/proc/tunderscore.nim:20]: @xflywind  missing foo causes CI failures

File: tests/proc/tunderscore.nim [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  proc ok(_, _: int) = discard
+  foo(4, 2)
+
+block:
+  proc ok(_: int, _: float) = discard
+  foo(1, 2.0)
+
+block:
+  proc ok(_: int, _: float, _: string) = discard
+  foo(1, 2.6, "5")
```
Comment (ringabout) [inline tests/proc/tunderscore.nim]:
```suggestion
block:
  proc ok(_, _: int) = discard
  ok(4, 2)

block:
  proc ok(_, _: int) = discard
  ok(4, 2)

block:
  proc ok(_: int, _: float) = discard
  ok(1, 2.0)

block:
  proc ok(_: int, _: float, _: string) = discard
  ok(1, 2.6, "5")
```

File: compiler/semtypes.nim [hunk 5]
```diff
@@ -40,12 +40,6 @@ const
   errNoGenericParamsAllowedForX = "no generic parameters allowed for $1"
   errInOutFlagNotExtern = "the '$1' modifier can be used only with imported types"
 
-
-proc isDiscardUnderscoreV(v: PSym): bool =
```
Comment (timotheecour) [inline compiler/semtypes.nim]: i still don't like the duplication, instead just rename this to:
```nim
proc handleUnderscore(v: PSym): bool = ...
```

File: compiler/ccgtypes.nim [hunk 6]
```diff
@@ -54,6 +54,9 @@ proc mangleParamName(m: BModule; s: PSym): Rope =
   ## we cannot use 'sigConflicts' here since we have a BModule, not a BProc.
   ## Fortunately C's scoping rules are sane enough so that that doesn't
   ## cause any trouble.
+  if s.name.s == "_":
```
Comment (timotheecour) [inline compiler/ccgtypes.nim:57]: can there be something similar for js? or what prevents this?

Comment (ringabout) [reply compiler/ccgtypes.nim:57]: It just didn't work for JS VM I guess a known bug

Comment (ringabout) [reply compiler/ccgtypes.nim:57]: tunderscore.nim(9, 27) Error: cannot evaluate at compile time: x

File: tests/proc/tunderscore.nim [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    foo(1, 2)
+    doAssert x == 10
+
+  block:
+    proc foo(_: int, _: float, _: string) = x += 5
+
+    foo(1, 2, "5")
+    doAssert x == 15
+
```
Comment (timotheecour) [inline tests/proc/tunderscore.nim:57]: add an example with a template:
```suggestion
  block:
    template foo(_, _, a): int = a
    doAssert fo(1,2,3) == 3
```

(and IMO you can simplify the other examples using the same pattern as here)

File: compiler/ccgtypes.nim (hunk 6 above)
Comment (timotheecour) [reply compiler/ccgtypes.nim:57]: in this case we should leave #13443 open (with a note) until we can address this :)

Comment (ringabout) [reply compiler/ccgtypes.nim:57]: why, that's the bug of js vm, not related to my PR?

File: tests/proc/tunderscore.nim (hunk 7 above)
Comment (ringabout) [reply tests/proc/tunderscore.nim:57]: make sense

File: compiler/ccgtypes.nim (hunk 6 above)
Comment (timotheecour) [reply compiler/ccgtypes.nim:57]: fine, I'll just open a new bug for js+vm then

**Review by timotheecour — DISMISSED** [review]

File: compiler/ccgtypes.nim (hunk 6 above)
Comment (Clyybber) [reply compiler/ccgtypes.nim:57]: This shouldn't be handled in the backend IMO.

File: tests/proc/tunderscore.nim [hunk 8]
```diff
@@ -0,0 +1,64 @@
+discard """
+  targets: "c cpp js"
+"""
+proc test() =
+  block:
+    proc ok(_, _, a: int): int = a
+    doassert ok(4, 2, 5) == 5
+
```
Comment (timotheecour) [inline tests/proc/tunderscore.nim:8]:
```suggestion
  block:
    iterator fn(_, _: int, c: int): int = yield c
    doAssert toSeq(fn(1,2,3)) == @[3]
```

and add from std/sequtils import toSeq on top

File: compiler/ccgtypes.nim (hunk 6 above)
Comment (timotheecour) [reply compiler/ccgtypes.nim:57]: maybe, what do you suggest concretely?
(see also the items i added in future work)

Comment (metagn) [reply compiler/ccgtypes.nim:57]: Can't we just mangle `if sfGensym in s.flags`?

```suggestion
  if sfGensym in s.flags:
```

Comment (metagn) [inline compiler/ccgtypes.nim:57]:
```suggestion
  if sfGensym in s.flags:
```

**Comment by metagn** [comment]
This PR is needed to fix #13804 (which the checkboxes in the main post describe). There would only be a small patch in semtempl, namely putting these lines inside `if param.name.s != "_":`

https://github.com/nim-lang/Nim/blob/779b1cc5beefa6064cb640401f1969c95a96c205/compiler/semtempl.nim#L645-L646

If we don't want the backend mangling parameter names, we can whitelist the changes in `semtypes` to just templates.

As for the `isDiscardUnderscore` comments, we can just forward declare it in `semtypes`/move it there, or we can convert it to `template ignoreUnderscore(v: PSym, body: untyped)` which does `if v.name.s == "_": v.flags.incl(sfGensym) else: body`.

**Pull request closed without merging**